asapdiscovery/data/utils.py

```diff
@@ -580,107 +651,228 @@ def filter_molecules_dataframe(
-    include_flags = []
-    for _, row in mol_df.iterrows():
-        smiles = row["smiles"]
-        include_this_molecule = (
-            (retain_achiral and is_achiral(smiles))
-            or (retain_racemic and is_racemic(smiles))
-            or (retain_enantiopure and is_enantiopure(smiles))
-        )
-        include_flags.append(include_this_molecule)
-    mol_df = mol_df.loc[include_flags, :]
+    logging.debug("Filtering molecules dataframe")
+    ## Determine which molecules will be retained and add corresponding labels
+    ##  to the data frame
+    achiral_label = [is_achiral(smiles) for smiles in mol_df["smiles"]]
+    racemic_label = [is_racemic(smiles) for smiles in mol_df["smiles"]]
+    enantiopure_label = [is_enantiopure(smiles) for smiles in mol_df["smiles"]]
+    semiquant_label = [
+        is_semiquant(ic50) for ic50 in mol_df[f"{assay_name}: IC50 (µM)"]
+    ]
+    keep_idx = [
+        (retain_achiral and achiral_label[i])
+        or (retain_racemic and racemic_label[i])
+        or (retain_enantiopure and enantiopure_label[i])
+        or (retain_semiquantitative_data and semiquant_label[i])
+        for i in range(mol_df.shape[0])
+    ]
 
-    # Filter out semiquantitative data, if requested
-    if not retain_semiquantitative_data:
-        include_flags = []
-        for _, row in mol_df.iterrows():
-            try:
-                _ = float(row[f"{assay_name}: IC50 (µM)"])
-                include_flags.append(True)
-            except ValueError as e:
-                include_flags.append(False)
+    mol_df["achiral"] = achiral_label
+    mol_df["racemic"] = racemic_label
+    mol_df["enantiopure"] = enantiopure_label
+    mol_df["semiquant"] = semiquant_label
 
-        mol_df = mol_df.loc[include_flags, :]
-        logging.debug(
-            (
-                f"  dataframe contains {mol_df.shape} entries after removing "
-                "semiquantitative data"
-            )
-        )
+    mol_df = mol_df.loc[keep_idx, :]
+    logging.debug(
+        f"  dataframe contains {mol_df.shape[0]} entries after filtering"
+    )
 
     # Compute pIC50s and uncertainties from 95% CIs
-    # TODO: In future, we can provide CIs as well
+    IC50_series = []
+    IC50_stderr_series = []
+    IC50_lower_series = []
+    IC50_upper_series = []
     pIC50_series = []
     pIC50_stderr_series = []
+    pIC50_range_series = []
+    pIC50_lower_series = []
+    pIC50_upper_series = []
     for _, row in mol_df.iterrows():
-        pIC50 = row[f"{assay_name}: Avg pIC50"]  # string
         try:
-            IC50 = float(row[f"{assay_name}: IC50 (µM)"]) * 1e-6  # molar
-            IC50_lower = (
-                float(row[f"{assay_name}: IC50 CI (Lower) (µM)"]) * 1e-6
-            )  # molar
-            IC50_upper = (
-                float(row[f"{assay_name}: IC50 CI (Upper) (µM)"]) * 1e-6
-            )  # molar
-
-            pIC50 = -np.log10(IC50)
+            IC50 = float(row[f"{assay_name}: IC50 (µM)"])
```

Comment:
Let's convert IC50 to M (from µM) here, so that subsequent work stores `IC50` in units of M.